app/actions/user.ts

```diff
@@ -1,171 +1,161 @@
-import { providers } from "ethers";
+import { providers, ethers, BigNumber } from "ethers";
 import { Thunk } from "state";
 
-import { addresses } from "@klimadao/lib/constants";
 import {
   formatUnits,
-  trimStringDecimals,
   getContract,
+  createContractsObject,
+  getAllowance,
+  getSpendersAndTokens,
+  getTokenDecimals,
 } from "@klimadao/lib/utils";
 import {
-  setBalance,
-  setExerciseAllowance,
-  setStakeAllowance,
-  setWrapAllowance,
-  setDomains,
-} from "state/user";
+  Allowances,
+  AllowancesSpender,
+  AllowancesFormatted,
+} from "@klimadao/lib/types/allowances";
+import { setBalance, updateAllowances, setDomains } from "state/user";
 
 import { getKns, getEns } from "./utils";
 
+const assets = [
+  "bct",
+  "nct",
+  "mco2",
+  "ubo",
+  "nbo",
+  "usdc",
+  "klima",
+  "sklima",
+  "wsklima",
+  "pklima",
+] as const;
+type Asset = typeof assets[number];
+
+const spenders = [
+  "staking_helper",
+  "staking",
+  "wsklima",
+  "pklima_exercise",
+] as const;
+
+type TokenValue = {
+  [K in Asset]: BigNumber;
+};
+type TokenValueFormatted = {
+  [K in Asset]: string;
+};
+
+const getBalance = async (params: {
+  token: Asset;
+  contract: ethers.Contract;
+  address: string;
+}): Promise<TokenValue> => {
+  try {
+    const balance = await params.contract.balanceOf(params.address);
+    return {
+      [params.token]: balance,
```

Comment:
curious why you opt for an unformatted value in an object rather than just returning a formatted string?

Reply:
Is changed